packages/block-editor/src/components/date-format-picker/index.js

```diff
@@ -0,0 +1,170 @@
+					if ( value === 'default' ) {
+						onChange( null );
+					} else if ( value === 'custom' ) {
+						onChange( defaultFormat );
+					}
+				} }
+			>
+				<ToggleGroupControlOption
+					value="default"
+					label={ __( 'Default' ) }
+				/>
+				<ToggleGroupControlOption
+					value="custom"
+					label={ __( 'Custom' ) }
+				/>
+			</ToggleGroupControl>
+			{ format && (
+				<CustomControls format={ format } onChange={ onChange } />
+			) }
+		</fieldset>
+	);
+}
+
+function CustomControls( { format, onChange } ) {
+	// Suggest a short format, medium format, long format, and a standardised
+	// (YYYY-MM-DD) format. The short, medium, and long formats are localised as
+	// different languages have different ways of writing these. For example, 'F
+	// j, Y' (April 20, 2022) in American English (en_US) is 'j. F Y' (20. April
+	// 2022) in German (de). The resultant array is de-duplicated as some
+	// languages will use the same format string for short, medium, and long
+	// formats.
+	const suggestedFormats = uniq( [
+		'Y-m-d',
+		_x( 'n/j/Y', 'short date format' ),
+		_x( 'n/j/Y g:i A', 'short date format with time' ),
+		_x( 'M j, Y', 'medium date format' ),
+		_x( 'M j, Y g:i A', 'medium date format with time' ),
+		_x( 'F j, Y', 'long date format' ),
+	] );
+
+	const suggestedOptions = suggestedFormats.map(
+		( suggestedFormat, index ) => ( {
+			key: `suggested-${ index }`,
+			name: dateI18n( suggestedFormat, EXAMPLE_DATE ),
+			format: suggestedFormat,
+		} )
+	);
+	const otherOption = {
+		key: 'other',
+		name: __( 'Other' ),
+		className:
+			'block-editor-date-format-picker__custom-format-select-control__other-option',
+		__experimentalHint: __( 'Choose your own date string' ),
+	};
+
+	const [ isOther, setIsOther ] = useState(
+		() => !! format && ! suggestedFormats.includes( format )
+	);
+
+	return (
+		<>
+			<BaseControl className="block-editor-date-format-picker__custom-format-select-control">
+				<CustomSelectControl
+					label={ __( 'Choose a format' ) }
+					options={ [ ...suggestedOptions, otherOption ] }
+					value={
+						isOther
+							? otherOption
+							: suggestedOptions.find(
+									( option ) => option.format === format
+							  ) ?? otherOption
+					}
+					onChange={ ( { selectedItem } ) => {
+						if ( selectedItem === otherOption ) {
+							setIsOther( true );
+						} else {
+							setIsOther( false );
+							onChange( selectedItem.format );
+						}
+					} }
+				/>
+			</BaseControl>
+			{ isOther && (
+				<TextControl
+					label={ __( 'Format string' ) }
+					hideLabelFromVision
+					help={ createInterpolateElement(
+						__(
+							'Enter a date or time <Link>format string</Link>.'
+						),
+						{
+							Link: (
+								<ExternalLink
+									href={ __(
+										'https://wordpress.org/support/article/formatting-date-and-time/'
+									) }
+								/>
+							),
+						}
+					) }
+					value={ format }
+					onChange={ ( value ) => onChange( value ) }
+				/>
```

Comment:
Would be good to revise the use of the word 'string' here as well, IMO.

Reply:
Hm I want to use the same nomenclature that we use in [the docs](https://wordpress.org/support/article/formatting-date-and-time/), though.

> Notice the string of characters next to each Date and Time formatting in screenshot. This string is called a format string. Each letter represents specific part of Date or Time.

Reply:
This documentation also straight away mentions PHP functions and the PHP programming language. I don't think most end users should be exposed to this kind of stuff, so my opinion is that the docs are wrong here too. They seem like technical docs.

Reply:
True, the docs could use a bit of work. It's what we link to from Options → General. I want to be consistent there.

To be honest this is a very technical feature and I don't think there's any point shying away from that. Maybe one day we will build a drag-and-drop style interface like @andrewserong suggests and hide the technical nature of it, but until then I think we just have to wear it. (And hide it in a dropdown where only power users will look, which we've done 😀)